### PR TITLE
KEYCLOAK-18069 Migration of client policies JSON from Keycloak 13

### PR DESCRIPTION
**Pull request by mposolda**

### CODE REVIEW

File: core/src/main/java/org/keycloak/representations/idm/RealmRepresentation.java [hunk 1]
```diff
@@ -144,8 +152,11 @@
 
     // Client Policies/Profiles
 
-    protected ClientProfilesRepresentation clientProfiles;
-    protected ClientPoliciesRepresentation clientPolicies;
+    @JsonProperty("clientProfiles")
+    protected JsonNode clientProfiles;
+
+    @JsonProperty("clientPolicies")
+    protected JsonNode clientPolicies;
```
Comment (vmuzikar) [inline core/src/main/java/org/keycloak/representations/idm/RealmRepresentation.java:159]: I wonder whether we should have an explicit setter to be used by the Jackson mapper. This way we don't verify at this level that the provided profiles/policies are an actual and valid representation.

**Review by vmuzikar — COMMENTED** [review]
TBH, this approach seems to me little bit like a workaround. Wouldn't be a cleaner solution just to remove all policies/profiles during the migration from incompatible versions? As you mentioned, it was a preview, so it should be ok.

**Comment by mposolda** [comment]
@vmuzikar Yes, it would be. Thanks for consider that! I was thinking about that and in fact, I've did that in the recent version of this PR after some more testing and adding the additional test for admin console client policies endpoint. 

However it is not sufficient as the JSON with realm representation is parsed earlier than the migration is triggered. This is why I added this workaround to RealmRepresentation. Since we have a bit more time before code freeze, I consider do some more proper solution and add another method on the Migration interface to have a chance to be triggered before realm import, earlier than the JSON parsing has a chance to occured. In fact, it can just use "default" empty implementation on all interfaces earlier than MigrateTo14_0_0, so there should not be much changes needed . WDYT?

**Comment by vmuzikar** [comment]
> However it is not sufficient as the JSON with realm representation is parsed earlier than the migration is triggered. This is why I added this workaround to RealmRepresentation. Since we have a bit more time before code freeze, I consider do some more proper solution and add another method on the Migration interface to have a chance to be triggered before realm import, earlier than the JSON parsing has a chance to occured. In fact, it can just use "default" empty implementation on all interfaces earlier than MigrateTo14_0_0, so there should not be much changes needed . WDYT?

+1000
It's be IMHO the cleanest and even most future proof solution.

**Comment by mposolda** [comment]
> > However it is not sufficient as the JSON with realm representation is parsed earlier than the migration is triggered. This is why I added this workaround to RealmRepresentation. Since we have a bit more time before code freeze, I consider do some more proper solution and add another method on the Migration interface to have a chance to be triggered before realm import, earlier than the JSON parsing has a chance to occured. In fact, it can just use "default" empty implementation on all interfaces earlier than MigrateTo14_0_0, so there should not be much changes needed . WDYT?
> 
> +1000
> It's be IMHO the cleanest and even most future proof solution.

Thanks for the confirm. I will try to do it if there is a time before release - in the end it may not be so easy to do. As it will probably require some fasterxml/jackson tweaking and possibly inject some custom serializer/deserializer at the jackson level. This is due the fact that for example admin REST endpoint for import realm already accepts RealmRepresentation and hence the deserialization needs to be triggered earlier in the request by the RestEasy Jackson provider, so this provider needs to be already aware of those "migrators" somehow.

**Review by vmuzikar — APPROVED** [review]
Based on internal discussion, I'm approving the current approach. It'll be improved in a [follow-up task](https://issues.redhat.com/browse/KEYCLOAK-18407).

**Pull request merged**